src/libutils/oglapphelpers/msl.mm

```diff
@@ -0,0 +1,502 @@
+    
+    int uniformId = 1;
+    for (unsigned idx = 0; idx < maxUniforms; ++idx)
+    {
+        GpuShaderDesc::UniformData data;
+        m_shaderDesc->getUniform(idx, data);
+        switch(data.m_type)
+        {
+            case UNIFORM_BOOL:
+            case UNIFORM_DOUBLE:
+            case UNIFORM_FLOAT3:
+                // Part of uniform data buffer
+            break;
+                
+            case UNIFORM_VECTOR_INT:
+            {
+                int size = data.m_vectorInt.m_getSize();
+                const int dummyInt = 123456789;
+                const int* v = size == 0 ? &dummyInt : data.m_vectorInt.m_getVector();
+                size = size == 0 ? sizeof(int) : size * sizeof(int);
+                [renderCmdEncoder setFragmentBytes:v length:size * sizeof(int) atIndex:uniformId++];
+            }
+            break;
+                
+            case UNIFORM_VECTOR_FLOAT:
+            {
+                int size = data.m_vectorFloat.m_getSize();
+                const float dummyFloat = 123456789.0f;
+                const float* v = size == 0 ? &dummyFloat : data.m_vectorFloat.m_getVector();
+                size = size == 0 ? sizeof(float) : size * sizeof(float);
+                [renderCmdEncoder setFragmentBytes:v length:size atIndex:uniformId++];
+            }
+            break;
+                
+            case UNIFORM_UNKNOWN:
+                throw Exception("Unknown uniform type.");
+        };
+    }
+}
+
+bool MetalBuilder::buildPipelineStateObject(const std::string & clientShaderProgram)
+{
+    NSString* shaderSrc = [NSString stringWithUTF8String:clientShaderProgram.c_str()];
+    
+    NSError* error = nil;
+    MTLCompileOptions* options = [MTLCompileOptions new];
+    [options setLanguageVersion:MTLLanguageVersion2_0];
+    [options setFastMathEnabled:NO];
+    m_library = [[m_device newLibraryWithSource:shaderSrc options:options error:&error] autorelease];
+    
+    id<MTLFunction> vertexShader = [[m_library newFunctionWithName:@"ColorCorrectionVS"] autorelease];
+    id<MTLFunction> pixelShader  = [[m_library newFunctionWithName:@"ColorCorrectionPS"] autorelease];
+    
+    MTLRenderPipelineDescriptor* renderPipelineDesc = [[MTLRenderPipelineDescriptor new] autorelease];
+    [renderPipelineDesc setVertexFunction:vertexShader];
+    [renderPipelineDesc setFragmentFunction:pixelShader];
+    [renderPipelineDesc setVertexDescriptor:nil];
+    [renderPipelineDesc setRasterizationEnabled:YES];
+    [renderPipelineDesc setInputPrimitiveTopology:MTLPrimitiveTopologyClassTriangle];
+    [[renderPipelineDesc colorAttachments][0] setPixelFormat:MTLPixelFormatRGBA32Float];
+    m_PSO = [[m_device newRenderPipelineStateWithDescriptor:renderPipelineDesc error:&error] autorelease];
+    
+    if(error == nil)
+        return true;
+    
+    return false;
+}
+
+void MetalBuilder::triggerProgrammaticCaptureScope()
+{
+    if (@available(macOS 10.15, *)) {
```

Comment:
we actually don't need `triggerProgrammaticCaptureScope` and `stopProgrammaticCaptureScope` to be part of msl.mm. They are only for debug purposes. We can remove them.

Reply:
I would rather keep these and add a message that triggering capture is only supported on MacOS 10.15 and above. Since this is not a core functionality but more for debugging, that should be ok.